### PR TITLE
feat(uninstall): delete inventory data

### PR DESCRIPTION
**Pull request by Rom1-B**
Adds an option to delete inventory data:
- unlock fields
- native agent
- elements linked to the agent if the `glpiinventory` plugin is active

ref: !25150

### CODE REVIEW

File: inc/uninstall.class.php [hunk 1]
```diff
@@ -298,6 +298,10 @@ static function uninstall($type, $model_id, $tab_ids, $location) {
             }
          }
 
+         if ($model->fields['raz_glpiinventory'] == 1) {
+            self::deleteGlpiInventoryLink($type, $id);
```
Comment (cedric-anne) [inline inc/uninstall.class.php:302]:
```suggestion
            self::deleteGlpiInventoryLink($item);
```

File: inc/uninstall.class.php [hunk 2]
```diff
@@ -469,6 +473,46 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
       }
    }
 
+   /**
+    * Function to remove GLPI Inventory informations for an asset
+    *
+    * @param $itemtype the asset type
+    * @param $items_id the asset's ID in GLPI
```
Comment (cedric-anne) [inline inc/uninstall.class.php:480]:
```suggestion
    * @param CommonDBTM $item
```

File: inc/uninstall.class.php [hunk 3]
```diff
@@ -469,6 +473,46 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
...
+    *
+    * @return nothing
+   **/
+   static function deleteGlpiInventoryLink($itemtype, $items_id) {
```
Comment (cedric-anne) [inline inc/uninstall.class.php:484]:
```suggestion
   static function deleteGlpiInventoryLink(CommonDBTM $item) {
```

File: inc/uninstall.class.php [hunk 4]
```diff
@@ -469,6 +473,46 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
...
+
+      $plug = new Plugin();
+      if ($plug->isActivated('glpiinventory') && function_exists('plugin_pre_item_purge_glpiinventory')) {
+         $item = new $itemtype();
+         $item->getFromDB($items_id);
+
```
Comment (cedric-anne) [inline inc/uninstall.class.php:490]:
```suggestion
```

File: inc/uninstall.class.php [hunk 5]
```diff
@@ -469,6 +473,46 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
...
+         // let glpi-inventory to clean item if needed (agent / collect etc ..)
+         plugin_pre_item_purge_glpiinventory($item);
+      } else {
+         self::purgeAgents($itemtype, $items_id);
```
Comment (cedric-anne) [inline inc/uninstall.class.php:494]: No need to split this into many methods.
```suggestion
         $agent = new Agent();
         $agent->deleteByCriteria([
            'itemtype' => $item->getType(),
            'items_id' => $item->getID(),
         ]);
```

File: inc/uninstall.class.php [hunk 6]
```diff
@@ -469,6 +473,46 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
...
+      }
+
+      // purge lock manually because related computer is not purged
+      self::purgeLockedField($itemtype, $items_id);
```
Comment (cedric-anne) [inline inc/uninstall.class.php:498]: There is already a method for this.

```suggestion
     $lockedfield = new Lockedfield();
     if ($lockedfield->isHandled($item)) {
         $lockedfield->itemDeleted();
     }
```

File: inc/uninstall.class.php [hunk 7]
```diff
@@ -469,6 +473,46 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
...
+      ]);
+   }
+
+   static function purgeAgents($itemtype, $items_id) {
+      $agent = new Agent();
+      $agent->deleteByCriteria([
+         'itemtype' => $itemtype,
+         'items_id' => $items_id
+      ]);
+   }
```
Comment (cedric-anne) [inline inc/uninstall.class.php:515]:
```suggestion
```

File: setup.php [hunk 8]
```diff
@@ -28,7 +28,7 @@
  * -------------------------------------------------------------------------
  */
 
-define ('PLUGIN_UNINSTALL_VERSION', '2.8.0');
+define ('PLUGIN_UNINSTALL_VERSION', '2.8.1');
```
Comment (cedric-anne) [inline setup.php:31]: Will be do during release process.

```suggestion
define ('PLUGIN_UNINSTALL_VERSION', '2.8.0');
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: inc/uninstall.class.php (hunk 1 above)
Comment (cedric-anne) [reply inc/uninstall.class.php:302]: You missed this one.

File: inc/model.class.php [hunk 9]
```diff
@@ -376,6 +376,12 @@ function showPartFormUninstall() {
       Dropdown::showYesNo("raz_antivirus",
                           (isset($this->fields["raz_antivirus"])
                            ? $this->fields["raz_antivirus"] : 1));
+      echo "</td>";
+      echo "<td>" . __('Delete inventory data (Lock, Agent, Collect ...)', 'uninstall') . "</td>";
+      echo "<td>";
+      Dropdown::showYesNo("raz_glpiinventory",
+                           (isset($this->fields["raz_glpiinventory"])
+                           ? $this->fields["raz_glpiinventory"] : 0), -1, ['width' => '100%']);
```
Comment (cedric-anne) [inline inc/model.class.php:384]: This is not define for other fields, so I guess it should not be for this one.
```suggestion
                           ? $this->fields["raz_glpiinventory"] : 0), -1);
```

**Comment by stonebuzz** [comment]
unless I am wrong, several elements are missing from ```relation.constant.php```



We need (at least) : 

```php
'glpi_monitors'    => 'computers_id',
'glpi_networkports'    => ['items_id', 'itemtype'],
'glpi_networkname'    => ['items_id', 'itemtype'],
'glpi_ipaddresses'    => ['mainitems_id', 'mainitemtype'],
'glpi_items_operatingsystems'    => ['items_id', 'itemtype'],
'glpi_items_disks'    => ['items_id', 'itemtype'],
'glpi_items_devicebatteries'    => ['items_id', 'itemtype'],
//all related item like glpi_items_devicexxxxxxx
```

**Comment by Rom1-B** [comment]
> unless I am wrong, several elements are missing from `relation.constant.php`
> 
> image
> 
> We need (at least) :
> 
> ```
> 'glpi_monitors'                             => 'computers_id',
> 'glpi_networkports'                      => ['items_id', 'itemtype'],
> 'glpi_networkname'                     => ['items_id', 'itemtype'],
> 'glpi_ipaddresses'                       => ['mainitems_id', 'mainitemtype'],
> 'glpi_items_operatingsystems'    => ['items_id', 'itemtype'],
> 'glpi_items_disks'                        => ['items_id', 'itemtype'],
> 'glpi_items_devicebatteries'        => ['items_id', 'itemtype'],
> //all related item like glpi_items_devicexxxxxxx
> ```

The link seems to be via '_glpi_computers_items'.

**Comment by stonebuzz** [comment]
```_glpi_computers_items``` handle only

```$CFG_GLPI["directconnect_types"]          = ['Monitor', 'Peripheral', 'Phone', 'Printer'];```

**Review by trasher — APPROVED** [review]

File: inc/uninstall.class.php [hunk 10]
```diff
@@ -469,6 +473,70 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
...
+                  $tablename,
+                  [
+                     'is_deleted' => 0
+                  ],
+                  [
+                     $field => $item->getID(),
+                     'is_dynamic' => 1
+                  ]
+               );
+            }
```
Comment (cedric-anne) [inline inc/uninstall.class.php:535]: Technically, you may have an array containing 2 distinct foreign keys, for instance
```php
    'glpi_groups' => [
        'glpi_certificates'          => [
            'groups_id_tech',
            'groups_id',
        ],
```
, also, you cannot be sure that `itemtype` and `items_id` are declared in the correct order.


```suggestion
            if (
               is_array($field)
               && count($itemtype_match = preg_grep('/itemtype/', $field)) === 1
               && count($items_id_match = preg_grep('/items_id/', $field)) === 1
            ) {
               $DB->update(
                  $tablename,
                  [
                     'is_deleted' => 0
                  ],
                  [
                     $items_id_match[0] => $item->getID(),
                     $itemtype_match[1] => $item->getType(),
                     'is_dynamic' => 1
                  ]
               );
            } else {
               if (!is_array($field)) {
                  $field = [$field];
               }
               foreach ($field as $f) {
                  $DB->update(
                     $tablename,
                     [
                        'is_deleted' => 0
                     ],
                     [
                        $f => $item->getID(),
                        'is_dynamic' => 1
                     ]
                  );
               }
            }
```

File: inc/uninstall.class.php [hunk 11]
```diff
@@ -469,6 +473,70 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
...
+      $lockedfield = new Lockedfield();
+      if ($lockedfield->isHandled($item)) {
+         $lockedfield->itemDeleted();
+      }
+
+      // unlock item relations
+      global $DB;
+      $RELATION = getDbRelations();
+      if (isset($RELATION[$item->getTable()])) {
+         foreach ($RELATION[$item->getTable()] as $tablename => $field) {
```
Comment (cedric-anne) [inline inc/uninstall.class.php:508]: You should verify if related items have a `is_dynamic` property (can be done using `CommonDBTM::canBeDynamic()`.

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Comment by cedric-anne** [comment]
Will be strongly impacted by https://github.com/glpi-project/glpi/pull/13018

**Comment by cedric-anne** [comment]
@stonebuzz 

Can you check this PR ?

**Review by cedric-anne — APPROVED** [review]

File: inc/uninstall.class.php [hunk 12]
```diff
@@ -500,6 +503,141 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
...
+                     [
+                        'is_deleted' => 0,
+                        'is_dynamic' => 0
+                     ],
+                     [
+                        $items_id_field => $item->getID(),
+                        $itemtype_field => $item->getType(),
+                        'is_dynamic' => 1
+                     ]
+                  );
```
Comment (stonebuzz) [inline inc/uninstall.class.php:610]: SQL error trigger here

```shell
[2023-06-21 08:20:34] glpisqllog.ERROR: DBmysql::query() in /home/stanislas/TECLIB/DEV/GLPI/10.0-bugfixes/src/DBmysql.php line 379
  *** MySQL query error:
  SQL: UPDATE `glpi_ipaddresses` SET `is_deleted` = '0', `is_dynamic` = '0' WHERE (Computer) AND `is_dynamic` = '1'
  Error: Unknown column 'Computer' in 'where clause'
  Backtrace :
  src/DBmysql.php:1461                               
  plugins/uninstall/inc/uninstall.class.php:608      DBmysql->update()
  plugins/uninstall/inc/uninstall.class.php:123      PluginUninstallUninstall::deleteGlpiInventoryLink()
  plugins/uninstall/inc/uninstall.class.php:333      PluginUninstallUninstall::doOneUninstall()
  plugins/uninstall/front/action.php:76              PluginUninstallUninstall::uninstall()
  {"user":"2@stanislas-asus-desktop"} 
```

File: inc/uninstall.class.php [hunk 13]
```diff
@@ -500,6 +503,141 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
...
+         $DB->update(
+            Computer::getTable(),
+            ['is_dynamic' => false],
+            ['id' => $item->getID()]
+         );
+         //reload
+         $item->getFromDB($item->getID());
+      }
+   }
+
```
Comment (stonebuzz) [inline inc/uninstall.class.php:640]: Some data / link are not purged

image

**Review by stonebuzz — CHANGES_REQUESTED** [review]

**Review by stonebuzz — APPROVED** [review]

File: inc/uninstall.class.php [hunk 14]
```diff
@@ -510,7 +511,7 @@ static function deleteFusionInventoryLink($itemtype, $items_id) {
     *
     * @return nothing
    **/
-   static function deleteGlpiInventoryLink($item) {
+   static function deleteGlpiInventoryLink(&$item) {
```
Comment (cedric-anne) [inline inc/uninstall.class.php:514]: Objects are always passed by reference.
```suggestion
   static function deleteGlpiInventoryLink($item) {
```

**Pull request merged**